### PR TITLE
Emit warning for broken object

### LINKED ISSUE #5080: KeyError: '__wrapped__' with from xxx import yyy
# Checklist

- [X] Include a detailed description of the bug or suggestion
- [ ] `pip list` of the virtual environment you are using
- [X] pytest and operating system versions
- [X] Minimal example if possible

# Description

I implemented `tox` and `pytest` for automated testing.  When I run the [MWE](https://github.com/ZaydH/pytest_bug), I get a `KeyError: '__wrapped__'` when I perform a `from xxx import yyy`.  

In the example I included, I do not get the KeyError if I comment out the line: 

...
from sty import fg
...


# Pip List

No virtual environment used.

# Version Info

*Python*: 3.6.5 & 3.7.1
*PyTest*: 4.4.0
*OS*: MacOS Mojave 10.14.4

# MWE

I created a very simple [repo](https://github.com/ZaydH/pytest_bug).  If I clone the repo and call either `tox` or `pytest` in the root directory, I get the error.

# Example Error Message

```
➜  pytest_bug git:(master) ✗ tox
GLOB sdist-make: /Users/zayd/repos/pytest_bug/setup.py
py36 recreate: /Users/zayd/repos/pytest_bug/.tox/py36
py36 installdeps: pytest, sty
py36 inst: /Users/zayd/repos/pytest_bug/.tox/.tmp/package/1/stratego-0.0.0.zip
py36 installed: atomicwrites==1.3.0,attrs==19.1.0,more-itertools==7.0.0,pluggy==0.9.0,py==1.8.0,pytest==4.4.0,six==1.12.0,stratego==0.0.0,sty==1.0.0b9
py36 run-test-pre: PYTHONHASHSEED='591797669'
py36 run-test: commands[0] | pytest
============================================================================================================== test session starts ==============================================================================================================
platform darwin -- Python 3.6.5, pytest-4.4.0, py-1.8.0, pluggy-0.9.0
cachedir: .tox/py36/.pytest_cache
rootdir: /Users/zayd/repos/pytest_bug, inifile: tox.ini
collected 0 items / 1 errors                                                                                                                                                                                                                    

==================================================================================================================== ERRORS =====================================================================================================================
_____________________________________________________________________________________________________ ERROR collecting stratego/printer.py ______________________________________________________________________________________________________
../../.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/doctest.py:933: in find
    self._find(tests, obj, name, module, source_lines, globs, {})
.tox/py36/lib/python3.6/site-packages/_pytest/doctest.py:406: in _find
    self, tests, obj, name, module, source_lines, globs, seen
../../.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/doctest.py:992: in _find
    if ((inspect.isroutine(inspect.unwrap(val))
.tox/py36/lib/python3.6/site-packages/_pytest/doctest.py:377: in _mock_aware_unwrap
    return real_unwrap(obj, stop=_is_mocked)
../../.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/inspect.py:512: in unwrap
    while _is_wrapper(func):
../../.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/inspect.py:506: in _is_wrapper
    return hasattr(f, '__wrapped__') and not stop(f)
E   KeyError: '__wrapped__'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================================================ 1 error in 0.25 seconds ============================================================================================================
ERROR: InvocationError for command /Users/zayd/repos/pytest_bug/.tox/py36/bin/pytest (exited with code 2)
____________________________________________________________________________________________________________________ summary ____________________________________________________________________________________________________________________
ERROR:   py36: commands failed
```

My apologies if this bug is obvious or I am doing something wrong.

### PR DESCRIPTION
**Pull request by Zac-HD**
Closes #5080.

@RonnyPfannschmidt - I'm unconvinced that we should do anything more than close the issue, but if you want a fix I think this is it.

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
this needs a test, and its absolutely necessary for people understand what the heck is wrong in certain cases

anything else is hiding issues/errors under a incomprehensible mumble

File: src/_pytest/doctest.py [hunk 1]
```diff
@@ -368,10 +370,21 @@ def _patch_unwrap_mock_aware():
...
+                        obj, stop=lambda obj: _is_mocked(obj) or stop(obj)
+                    )
+            except Exception as e:
+                warnings.warn(
+                    "Got %r when unwrapping %r.  This is usually caused "
+                    "by a violation of Python's object protocol; see e.g. "
+                    "https://github.com/pytest-dev/pytest/issues/5080" % (e, obj),
+                    PytestWarning,
+                )
+                raise
```
Comment (blueyed) [inline src/_pytest/doctest.py]: How about making the code more readable in the first place?
```python
if stop is None:
    stop = _is_mocked
else:
    stop = lambda obj: _is_mocked(obj) or stop(obj)
return real_unwrap(obj, stop=stop)
```

```python
if stop is None:
    stop = _is_mocked
else:
    stop = lambda obj: _is_mocked(obj) or stop(obj)
return real_unwrap(obj, stop=stop)
```

**Comment by blueyed** [comment]
@RonnyPfannschmidt 
I agree that it needs a test, but only half of it appears to have been covered before already...

File: src/_pytest/doctest.py (hunk 1 above)
Comment (Zac-HD) [reply src/_pytest/doctest.py]: `flake8` error about assigning a lambda, and IMO defining a named function is less readable (and certainly longer).

**Comment by Zac-HD** [comment]
@blueyed - I've got a test, and it's definitely covering that branch locally, but for some reason codecov hasn't seen it.  Any idea what's going on?

**Comment by blueyed** [comment]
@Zac-HD 
Check the build logs, which call `coverage report -m` and see if it is missing there already.

The partial coverage of the decorator looks weird also: https://codecov.io/gh/pytest-dev/pytest/compare/4f597f011d3d45dc33230f6c3b0cf36f88976030...3ea718182938bb8da0d90bedb13e283890ea4340/diff#D1-1236

File: testing/test_doctest.py [hunk 2]
```diff
@@ -1224,3 +1226,19 @@ class Example(object):
...
+
+@pytest.mark.parametrize(
+    "stop", [None, lambda f: None, lambda f: False, lambda f: True]
+)
+def test_warning_on_unwrap_of_broken_object(stop):
+    bad_instance = Broken()
+    with _patch_unwrap_mock_aware():
+        with pytest.raises(KeyError):
+            with pytest.warns(pytest.PytestWarning):
+                inspect.unwrap(bad_instance, stop=stop)
```
Comment (blueyed) [inline testing/test_doctest.py]: Maybe something is causing coverage tracing to be reset here - like it happens with RuntimeErrors.

**Comment by Zac-HD** [comment]
Hmm.  https://travis-ci.org/pytest-dev/pytest/jobs/543079937#L5068 and https://travis-ci.org/pytest-dev/pytest/jobs/543079940#L417 don't list the line as missed (though they do have a missed 377-exit branch), so I suspect a problem where codecov is combining them.  I can try if/return;return instead of if/return;else/return...

**Comment by blueyed** [comment]
@Zac-HD 
The missing coverage here is due to the parametrized functions not being called, isn't it?
Pushed a fixup.

**Comment by blueyed** [comment]
squash-merging is disabled here.. :/  (https://github.com/pytest-dev/pytest/issues/4361)
I could have done it already, but was not sure if force-pushing to your branch is ok for you.

**Comment by Zac-HD** [comment]
I appreciate the thought - "please don't without warning" probably covers my view on force-pushes.  Anyway, I've squashed and force-pushed it now so once CI is done you should be able to merge it :smile:

**Comment by nicoddemus** [comment]
Thanks @Zac-HD!

I noticed we don't need the `real_unwrap = getattr(inspect, "unwrap", None)` line in `_patch_unwrap_mock_aware` as it was introduced in Python 3.4, but it is better to clean that up after this has been backported. 👍

**Comment by nicoddemus** [comment]
@blueyed want to take another look or can we merge this? 😁

**Comment by blueyed** [comment]
Thanks! :)

**Pull request merged**